File: python/src/kanbus/policy_context.py
"""Policy evaluation context and reporting types."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

from kanbus.models import IssueData, ProjectConfiguration


class PolicyOperation(str, Enum):
    """Type of operation triggering policy evaluation."""

    CREATE = "create"
    UPDATE = "update"
    CLOSE = "close"
    DELETE = "delete"
    VIEW = "view"
    LIST = "list"
    READY = "ready"


class GuidanceSeverity(str, Enum):
    """Guidance item severity."""

    WARNING = "warning"
    SUGGESTION = "suggestion"


@dataclass(frozen=True)
class GuidanceItem:
    """Non-blocking guidance emitted by policy evaluation."""

    severity: GuidanceSeverity
    message: str
    explanations: list[str]
    policy_file: str
    scenario: str
    step: str


@dataclass
class StatusTransition:
    """Status transition details.

    :param from_status: Status before the transition.
    :type from_status: str
    :param to_status: Status after the transition.
    :type to_status: str
    """

    from_status: str
    to_status: str


@dataclass
class PolicyContext:
    """Context provided to policy evaluation steps.

    Contains all information needed to evaluate policies against
    a proposed issue state change.

    :param current_issue: Current issue state on disk.
    :type current_issue: Optional[IssueData]
    :param proposed_issue: Proposed issue state after applying updates.
    :type proposed_issue: IssueData
    :param transition: Status transition details if status is changing.
    :type transition: Optional[StatusTransition]
    :param operation: Type of operation triggering evaluation.
    :type operation: PolicyOperation
    :param project_configuration: Project configuration.
    :type project_configuration: ProjectConfiguration
    :param all_issues: All issues in the project for aggregate checks.
    :type all_issues: list[IssueData]
    """

    current_issue: Optional[IssueData]
    proposed_issue: IssueData
    transition: Optional[StatusTransition]
    operation: PolicyOperation
    project_configuration: ProjectConfiguration
    all_issues: list[IssueData]

    @property
    def issue(self) -> IssueData:
        """Get the issue being evaluated (proposed state).

        :return: The proposed issue.
        :rtype: IssueData
        """
        return self.proposed_issue

    def is_transition(self) -> bool:
        """Check if this is a status transition.

        :return: True if status is changing.
        :rtype: bool
        """
        return self.transition is not None

    def is_transitioning_to(self, status: str) -> bool:
        """Check if transitioning to a specific status.

        :param status: Target status to check.
        :type status: str
        :return: True if transitioning to this status.
        :rtype: bool
        """
        return self.transition is not None and self.transition.to_status == status

    def is_transitioning_from(self, status: str) -> bool:
        """Check if transitioning from a specific status.

        :param status: Source status to check.
        :type status: str
        :return: True if transitioning from this status.
        :rtype: bool
        """
        return self.transition is not None and self.transition.from_status == status

    def child_issues(self) -> list[IssueData]:
        """Get child issues of the proposed issue.

        :return: List of child issues.
        :rtype: list[IssueData]
        """
        parent_id = self.proposed_issue.identifier
        return [issue for issue in self.all_issues if issue.parent == parent_id]

    def parent_issue(self) -> Optional[IssueData]:
        """Get parent issue of the proposed issue.

        :return: Parent issue if it exists.
        :rtype: Optional[IssueData]
        """
        if not self.proposed_issue.parent:
            return None
        parent_id = self.proposed_issue.parent
        return next(
            (issue for issue in self.all_issues if issue.identifier == parent_id),
            None,
        )


class PolicyViolationError(RuntimeError):
    """Raised when a policy is violated.

    :param policy_file: Path to the policy file.
    :type policy_file: str
    :param scenario: Scenario name that failed.
    :type scenario: str
    :param failed_step: The specific step that failed.
    :type failed_step: str
    :param message: Human-readable explanation.
    :type message: str
    :param issue_id: Issue ID being evaluated.
    :type issue_id: str
    """

    def __init__(
        self,
        policy_file: str,
        scenario: str,
        failed_step: str,
        message: str,
        issue_id: str,
        explanations: list[str] | None = None,
        guidance_items: list[GuidanceItem] | None = None,
    ) -> None:
        """Initialize policy violation error."""
        self.policy_file = policy_file
        self.scenario = scenario
        self.failed_step = failed_step
        self.message = message
        self.issue_id = issue_id
        self.explanations = explanations or []
        self.guidance_items = guidance_items or []

        lines = [
            f"policy violation in {policy_file} for issue {issue_id}",
            f"  Scenario: {scenario}",
            f"  Failed: {failed_step}",
            f"  {message}",
        ]
        for explanation in self.explanations:
            lines.append(f"  Explanation: {explanation}")

        ordered_guidance = sorted(
            self.guidance_items,
            key=lambda item: (
                0 if item.severity == GuidanceSeverity.WARNING else 1,
                item.policy_file,
                item.scenario,
                item.step,
                item.message,
            ),
        )
        if ordered_guidance:
            lines.append("  Guidance:")
            for item in ordered_guidance:
                prefix = (
                    "GUIDANCE WARNING"
                    if item.severity == GuidanceSeverity.WARNING
                    else "GUIDANCE SUGGESTION"
                )
                lines.append(f"    {prefix}: {item.message}")
                for explanation in item.explanations:
                    lines.append(f"      Explanation: {explanation}")
        super().__init__("\n".join(lines))


@dataclass
class PolicyEvaluationReport:
    """Result of policy evaluation in enforcement or guidance mode."""

    violations: list[PolicyViolationError] = field(default_factory=list)
    guidance_items: list[GuidanceItem] = field(default_factory=list)

File: rust/src/policy_context.rs
//! Policy evaluation context.

use crate::models::{IssueData, ProjectConfiguration};

/// Type of operation triggering policy evaluation.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PolicyOperation {
    /// Issue is being created.
    Create,
    /// Issue is being updated.
    Update,
    /// Issue is being closed.
    Close,
    /// Issue is being deleted.
    Delete,
    /// Issue is being viewed.
    View,
    /// Issues are being listed.
    List,
    /// Ready issues are being listed.
    Ready,
}

/// Status transition details.
#[derive(Debug, Clone)]
pub struct StatusTransition {
    /// Status before the transition.
    pub from: String,
    /// Status after the transition.
    pub to: String,
}

/// Context provided to policy evaluation steps.
///
/// Contains all information needed to evaluate policies against
/// a proposed issue state change.
#[derive(Debug, Clone)]
pub struct PolicyContext {
    /// Current issue state on disk.
    pub current_issue: Option<IssueData>,
    /// Proposed issue state after applying updates.
    pub proposed_issue: IssueData,
    /// Status transition details if status is changing.
    pub transition: Option<StatusTransition>,
    /// Type of operation triggering evaluation.
    pub operation: PolicyOperation,
    /// Project configuration.
    pub project_configuration: ProjectConfiguration,
    /// All issues in the project for aggregate checks.
    pub all_issues: Vec<IssueData>,
}

impl PolicyContext {
    /// Get the issue being evaluated (proposed state).
    pub fn issue(&self) -> &IssueData {
        &self.proposed_issue
    }

    /// Check if this is a status transition.
    pub fn is_transition(&self) -> bool {
        self.transition.is_some()
    }

    /// Check if transitioning to a specific status.
    pub fn is_transitioning_to(&self, status: &str) -> bool {
        self.transition
            .as_ref()
            .map(|t| t.to == status)
            .unwrap_or(false)
    }

    /// Check if transitioning from a specific status.
    pub fn is_transitioning_from(&self, status: &str) -> bool {
        self.transition
            .as_ref()
            .map(|t| t.from == status)
            .unwrap_or(false)
    }

    /// Get child issues of the proposed issue.
    pub fn child_issues(&self) -> Vec<&IssueData> {
        let parent_id = &self.proposed_issue.identifier;
        self.all_issues
            .iter()
            .filter(|issue| issue.parent.as_ref() == Some(parent_id))
            .collect()
    }

    /// Get parent issue of the proposed issue.
    pub fn parent_issue(&self) -> Option<&IssueData> {
        let parent_id = self.proposed_issue.parent.as_ref()?;
        self.all_issues
            .iter()
            .find(|issue| &issue.identifier == parent_id)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{PriorityDefinition, StatusDefinition};
    use chrono::Utc;
    use std::collections::BTreeMap;

    fn sample_issue(identifier: &str, parent: Option<&str>) -> IssueData {
        IssueData {
            identifier: identifier.to_string(),
            title: format!("Issue {identifier}"),
            description: String::new(),
            issue_type: "task".to_string(),
            status: "open".to_string(),
            priority: 2,
            assignee: None,
            creator: None,
            parent: parent.map(std::string::ToString::to_string),
            labels: Vec::new(),
            dependencies: Vec::new(),
            comments: Vec::new(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
            closed_at: None,
            custom: BTreeMap::new(),
        }
    }

    fn sample_project_configuration() -> ProjectConfiguration {
        let mut workflows = BTreeMap::new();
        workflows.insert("default".to_string(), BTreeMap::new());

        let mut priorities = BTreeMap::new();
        priorities.insert(
            2u8,
            PriorityDefinition {
                name: "medium".to_string(),
                color: None,
            },
        );

        ProjectConfiguration {
            project_directory: "project".to_string(),
            virtual_projects: BTreeMap::new(),
            new_issue_project: None,
            ignore_paths: Vec::new(),
            console_port: None,
            project_key: "kanbus".to_string(),
            project_management_template: None,
            hierarchy: vec!["task".to_string()],
            types: vec!["task".to_string()],
            workflows,
            transition_labels: BTreeMap::new(),
            initial_status: "open".to_string(),
            priorities,
            default_priority: 2,
            assignee: None,
            time_zone: None,
            statuses: vec![StatusDefinition {
                key: "open".to_string(),
                name: "Open".to_string(),
                category: "todo".to_string(),
                color: None,
                collapsed: false,
            }],
            categories: Vec::new(),
            sort_order: BTreeMap::new(),
            type_colors: BTreeMap::new(),
            beads_compatibility: false,
            wiki_directory: None,
            ai: None,
            jira: None,
            snyk: None,
            realtime: Default::default(),
            overlay: Default::default(),
            hooks: Default::default(),
        }
    }

    fn sample_context(transition: Option<StatusTransition>, parent: Option<&str>) -> PolicyContext {
        let proposed = sample_issue("kanbus-parent", parent);
        let child = sample_issue("kanbus-child", Some("kanbus-parent"));
        let other = sample_issue("kanbus-other", None);
        PolicyContext {
            current_issue: None,
            proposed_issue: proposed,
            transition,
            operation: PolicyOperation::Update,
            project_configuration: sample_project_configuration(),
            all_issues: vec![child, other],
        }
    }

    #[test]
    fn transition_helpers_cover_true_and_false_paths() {
        let with_transition = sample_context(
            Some(StatusTransition {
                from: "open".to_string(),
                to: "in_progress".to_string(),
            }),
            None,
        );
        assert!(with_transition.is_transition());
        assert!(with_transition.is_transitioning_to("in_progress"));
        assert!(with_transition.is_transitioning_from("open"));
        assert!(!with_transition.is_transitioning_to("closed"));
        assert!(!with_transition.is_transitioning_from("blocked"));

        let no_transition = sample_context(None, None);
        assert!(!no_transition.is_transition());
        assert!(!no_transition.is_transitioning_to("open"));
        assert!(!no_transition.is_transitioning_from("open"));
    }

    #[test]
    fn issue_parent_and_child_helpers_return_expected_records() {
        let no_parent = sample_context(None, None);
        let children = no_parent.child_issues();
        assert_eq!(children.len(), 1);
        assert_eq!(children[0].identifier, "kanbus-child");
        assert!(no_parent.parent_issue().is_none());
        assert_eq!(no_parent.issue().identifier, "kanbus-parent");

        let with_parent = sample_context(None, Some("kanbus-other"));
        let parent = with_parent.parent_issue().expect("parent should resolve");
        assert_eq!(parent.identifier, "kanbus-other");
    }
}
